apps/cowswap-frontend/src/common/hooks/useAutoFitText.ts

```diff
@@ -0,0 +1,128 @@
+import { DependencyList, RefObject, useLayoutEffect, useRef } from 'react'
+
+type FitMode = 'single' | 'multi'
+
+interface AutoFitOptions {
+  min?: number
+  max?: number
+  step?: number
+  mode?: FitMode
+  deps?: DependencyList
+}
+
+const EMPTY_DEPS: DependencyList = []
+
+interface NormalizedOptions {
+  minBound: number
+  maxBound: number
+  stepSize: number
+  shouldWarn: boolean
+}
+
+interface FitConfig {
+  node: HTMLElement
+  parent: HTMLElement
+  minBound: number
+  maxBound: number
+  stepSize: number
+  whiteSpace: 'nowrap' | 'normal'
+}
+
+function normalizeOptions(min: number, max: number, step: number): NormalizedOptions {
+  const stepSize = step > 0 ? step : 1
+  const swappedBounds = min > max
+
+  return {
+    minBound: swappedBounds ? max : min,
+    maxBound: swappedBounds ? min : max,
+    stepSize,
+    shouldWarn: swappedBounds || stepSize !== step,
+  }
+}
+
+function fitText({ node, parent, minBound, maxBound, stepSize, whiteSpace }: FitConfig): void {
+  const setSize = (val: number): void => {
+    node.style.fontSize = `${val}px`
+    node.style.whiteSpace = whiteSpace
+  }
+
+  const fits = (): boolean => {
+    const { width, height } = node.getBoundingClientRect()
+    const { width: pWidth, height: pHeight } = parent.getBoundingClientRect()
+    return width <= pWidth && height <= pHeight
+  }
+
+  let low = minBound
+  let high = maxBound
+  let best = minBound
+
+  while (low <= high) {
+    const mid = Math.floor((low + high) / 2)
+    setSize(mid)
+
+    if (fits()) {
+      best = mid
+      low = mid + 1
+    } else {
+      high = mid - 1
+    }
+  }
+
+  // Quantize downward so we never exceed the measured best-fit size
+  const quantized = Math.floor(best / stepSize) * stepSize
+  const clamped = Math.max(minBound, Math.min(quantized, best, maxBound))
+  setSize(clamped)
+}
+
+/**
+ * Lightweight auto-fit hook to scale text so it fits its parent box.
+ * It uses binary search over font sizes and ResizeObserver to re-apply on layout changes.
```

Comment:
Pardon my ignorance, but where is the binary search in this hook?
I probably missed something obvious, because I don't see it.